autorest/azure/auth/auth.go

```diff
@@ -0,0 +1,197 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to start device auth flow: %s", err)
+	}
+
+	fmt.Println(*deviceCode.Message)
+
+	token, err := adal.WaitForUserCompletion(oauthClient, deviceCode)
+	if err != nil {
+		return nil, fmt.Errorf("failed to finish device auth flow: %s", err)
+	}
+
+	spToken, err := adal.NewServicePrincipalTokenFromManualToken(*oauthConfig, options.ClientID, options.Environment.ResourceManagerEndpoint, *token)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get oauth token from device flow: %v", err)
+	}
+
+	return autorest.NewBearerAuthorizer(spToken), nil
+}
+
+func decodePkcs12(pkcs []byte, password string) (*x509.Certificate, *rsa.PrivateKey, error) {
+	privateKey, certificate, err := pkcs12.Decode(pkcs, password)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	rsaPrivateKey, isRsaKey := privateKey.(*rsa.PrivateKey)
+	if !isRsaKey {
+		return nil, nil, fmt.Errorf("PKCS#12 certificate must contain an RSA private key")
+	}
+
+	return certificate, rsaPrivateKey, nil
+}
+
+// AuthorizerOptions provides the options to get an authorizer
+type AuthorizerOptions struct {
+	TenantID            string
+	ClientID            string
+	ClientSecret        string
+	CertificatePath     string
+	CertificatePassword string
+	Environment         azure.Environment
+}
+
+// NewAuthorizerOptionsForClientCredentials return an AuthorizerOptions object configured to obtain an Authorizer through Client Credentials.
+func NewAuthorizerOptionsForClientCredentials(tenantID string, clientID string, clientSecret string, env azure.Environment) AuthorizerOptions {
+	return AuthorizerOptions{
+		TenantID:     tenantID,
+		ClientID:     clientID,
+		ClientSecret: clientSecret,
+		Environment:  env,
+	}
+}
+
+// NewAuthorizerOptionsForClientCertificate return an AuthorizerOptions object configured to obtain an Authorizer through client certificate.
+func NewAuthorizerOptionsForClientCertificate(tenantID string, clientID string, certificatePath string, certificatePassword string, env azure.Environment) AuthorizerOptions {
+	return AuthorizerOptions{
+		TenantID:            tenantID,
+		ClientID:            clientID,
+		CertificatePath:     certificatePath,
+		CertificatePassword: certificatePassword,
+		Environment:         env,
+	}
+}
+
+// NewAuthorizerOptionsFromEnvironment return an AuthorizerOptions configured from environment variables.
+func NewAuthorizerOptionsFromEnvironment() AuthorizerOptions {
+	options := AuthorizerOptions{}
+	options.TenantID = os.Getenv("AZURE_TENANT_ID")
+	options.ClientID = os.Getenv("AZURE_CLIENT_ID")
+	options.ClientSecret = os.Getenv("AZURE_CLIENT_SECRET")
+	options.CertificatePath = os.Getenv("AZURE_CERTIFICATE_PATH")
+	options.CertificatePassword = os.Getenv("AZURE_CERTIFICATE_PASSWORD")
+
+	envName := os.Getenv("AZURE_ENVIRONMENT")
+	env, err := azure.EnvironmentFromName(envName)
+
+	if err == nil {
+		options.Environment = env
+	} else {
+		options.Environment = azure.PublicCloud
```

Comment:
Defaulting like this seems dangerous.  What if I specified AZURE_ENVIRONMENT but misspelled the name?  I think this should be something like the following.
```golang
envName := os.Getenv("AZURE_ENVIRONMENT")
if envName == "" {
    options.Environment = azure.PublicCloud
} else {
    env, err := azure.EnvironmentFromName(envName)
    if err != nil {
        panic(err)
    }
    options.Environment = env
}
```

Reply:
Yes, good point